File: ur_control_client/src/example_movej.cpp
/**
 * @file example_movej.cpp
 * @brief MoveJ motion control example
 *
 * Demonstrates how to control UR robot using joint space motion (MoveJ)
 */

#include <chrono>
#include <rclcpp/rclcpp.hpp>
#include <thread>

#include "ur_control_client/ur_control_client.hpp"

using namespace std::chrono_literals;

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);

    auto client = std::make_shared<URControlClient>();

    // Spin in background thread
    std::thread spin_thread([client]() {
        rclcpp::spin(client);
    });

    // Wait for robot connection
    RCLCPP_INFO(client->get_logger(), "Waiting for robot connection...");
    while (rclcpp::ok() && !client->isConnected()) {
        std::this_thread::sleep_for(100ms);
    }

    if (!client->isConnected()) {
        RCLCPP_ERROR(client->get_logger(), "Failed to connect to robot");
        rclcpp::shutdown();
        spin_thread.join();
        return 1;
    }

    RCLCPP_INFO(client->get_logger(), "✅ Robot connected!");
    RCLCPP_INFO(client->get_logger(), "");
    RCLCPP_INFO(client->get_logger(), "========================================");
    RCLCPP_INFO(client->get_logger(), "  MoveJ Example: Joint Space Motion");
    RCLCPP_INFO(client->get_logger(), "========================================");
    RCLCPP_INFO(client->get_logger(), "");

    // Example 1: Move to home position
    RCLCPP_INFO(client->get_logger(), "Example 1: Moving to HOME position");
    std::vector<double> home = {0.0, -1.57, 1.57, -1.57, -1.57, 0.0};

    if (client->moveJ(home, 0.3, true)) {
        RCLCPP_INFO(client->get_logger(), "✅ Reached HOME position");
    } else {
        RCLCPP_ERROR(client->get_logger(), "❌ Failed to reach HOME position");
    }

    std::this_thread::sleep_for(1s);

    // Example 2: Move to target position
    RCLCPP_INFO(client->get_logger(), "");
    RCLCPP_INFO(client->get_logger(), "Example 2: Moving to TARGET position");
    std::vector<double> target = {0.5, -1.2, 1.0, -1.5, -1.57, 0.5};

    if (client->moveJ(target, 0.5, true)) {
        RCLCPP_INFO(client->get_logger(), "✅ Reached TARGET position");
    } else {
        RCLCPP_ERROR(client->get_logger(), "❌ Failed to reach TARGET position");
    }

    std::this_thread::sleep_for(1s);

    // Example 3: Return to home
    RCLCPP_INFO(client->get_logger(), "");
    RCLCPP_INFO(client->get_logger(), "Example 3: Returning to HOME");

    if (client->moveJ(home, 0.3, true)) {
        RCLCPP_INFO(client->get_logger(), "✅ Returned to HOME position");
    } else {
        RCLCPP_ERROR(client->get_logger(), "❌ Failed to return to HOME");
    }

    RCLCPP_INFO(client->get_logger(), "");
    RCLCPP_INFO(client->get_logger(), "========================================");
    RCLCPP_INFO(client->get_logger(), "  MoveJ Example Completed!");
    RCLCPP_INFO(client->get_logger(), "========================================");

    rclcpp::shutdown();
    spin_thread.join();

    return 0;
}

File: ur_control_client_py/ur_control_client_py/example_movej.py
#!/usr/bin/env python3
"""
MoveJ Motion Control Example (Python)
======================================
Demonstrates how to control UR robot using joint space motion (MoveJ)
"""

import asyncio

import rclpy
from rclpy.node import Node
from ur_control_client_py import URRobotClient


async def spin_node(node):
    """Spin ROS node in async loop"""
    while rclpy.ok():
        rclpy.spin_once(node, timeout_sec=0.1)
        await asyncio.sleep(0.01)


async def main_async():
    rclpy.init()

    # Create ROS2 node
    node = Node("ur_control_example_movej")

    # Create robot controller
    client = URRobotClient(node)

    # Start ROS spinning task
    spin_task = asyncio.create_task(spin_node(node))

    # Wait for robot connection
    node.get_logger().info("Waiting for robot connection and to be ready...")
    ready = await client.wait_robot_ready(timeout=5.0)

    if not ready:
        node.get_logger().error(
            "❌ Robot not ready (timeout waiting for joint/speed/tcp/io)"
        )
        spin_task.cancel()
        node.destroy_node()
        rclpy.shutdown()
        return

    node.get_logger().info("✅ Robot connected and ready!")
    node.get_logger().info("")
    node.get_logger().info("=" * 60)
    node.get_logger().info("  MoveJ Example: Joint Space Motion (Python)")
    node.get_logger().info("=" * 60)
    node.get_logger().info("")

    # Example 1: Move to home position
    node.get_logger().info("Example 1: Moving to HOME position")
    home = [0.0, -1.57, 1.57, -1.57, -1.57, 0.0]

    success, msg = await client.move_j(home, velocity=0.3)
    if success:
        node.get_logger().info("✅ Reached HOME position")
    else:
        node.get_logger().error(f"❌ Failed: {msg}")

    await client.wait(1.0)

    # Example 2: Move to target position
    node.get_logger().info("")
    node.get_logger().info("Example 2: Moving to TARGET position")
    target = [0.5, -1.2, 1.0, -1.5, -1.57, 0.5]

    success, msg = await client.move_j(target, velocity=0.5)
    if success:
        node.get_logger().info("✅ Reached TARGET position")
    else:
        node.get_logger().error(f"❌ Failed: {msg}")

    await client.wait(1.0)

    # Example 3: Return to home
    node.get_logger().info("")
    node.get_logger().info("Example 3: Returning to HOME")

    success, msg = await client.move_j(home, velocity=0.3)
    if success:
        node.get_logger().info("✅ Returned to HOME position")
    else:
        node.get_logger().error(f"❌ Failed: {msg}")

    node.get_logger().info("")
    node.get_logger().info("=" * 60)
    node.get_logger().info("  MoveJ Example Completed!")
    node.get_logger().info("=" * 60)

    spin_task.cancel()
    node.destroy_node()
    rclpy.shutdown()


def main():
    asyncio.run(main_async())


if __name__ == "__main__":
    main()
